[김민지] Almost completed

minzzzy opened the pull request:
Maybe complete mission😆

minzzzy committed:
feat: Add cart entity

minzzzy committed:
feat: Add Cart logic with test

javajigi reviewed (commented):
미션 진행하느라 수고했어요.

minzzzy committed:
feat: Add cart front-end in main page

- Create cart class

minzzzy committed:
refactor: Send Ajax request for adding a product in cart

- Add cartFetchManager

minzzzy committed:
feat: Add ResponseModel

minzzzy committed:
feat: Add CartAddDto to transmitting product id

minzzzy committed:
refactor: Refactor ApiCartController for adding product.

- Get user from session
- Use builder pattern User
- Add function in CartRepository
- Replace mapping relation in Cart

minzzzy committed:
chore: Remove unnecessary data

minzzzy committed:
chore: Remove unnecessary data

- Remove data related to Cart

minzzzy committed:
refactor: Refactor front-end to change product number in cart

minzzzy committed:
refactor: Refactor to add product in cart

- Add extra fields and functions in Cart entity
- Remove unnecessary field in Cart entity
- Replace file name from CartAddDto to CartDto and add product number
- Add cart session
- Refactor to add product in cart for keeping cart after login with acceptance test
- Remove CartRepository test
- Add Cart Service

minzzzy committed:
refactor: Refactor to add product in cart

- Add extra fields and functions in Cart entity
- Remove CartDto
- Change cart to cart id for setting cart session
- Change to add product by using path variable

minzzzy committed:
test: Add cart and cartService test

minzzzy committed:
feat: Show product list in cart and compute total products' price

- Create cart controller and remove '/cart' routing from home controller
- Create cartProductDto with test
- Add cartProduct template
- Add cartService unit test

minzzzy committed:
feat: Create computing delivery price logic with test

- Remove unnecessary field in cart
- Show delivery price and total price

minzzzy committed:
feat: Compute product price by product number and discount ratio

- Replace type of discount ratio from long to double
- Add compute discount ratio with test

minzzzy committed:
refactor: Refactor fetchManager

minzzzy committed:
refactor: Rename 'cart.js' to 'products.js'

minzzzy committed:
chore: Remove unused method

minzzzy committed:
refactor: Change Api url to request products in cart

minzzzy committed:
feat: Add product to cart with test

- Set cart session for keeping cart data after login
- Refactor cartController
- Delete cart acceptanc test
- Refactor cart product DTO to show products in cart page & test
- Add cart price DTO with test
- Add CartStaticValue to collect static values
- Move builder annotation in user
- Refactor sessionUtils
- Split javascript files and refactor them
-

minzzzy committed:
fix: Fix bugs to pass test of user

- Add already singup user
- Add passwordConfirm getter in userDTO
- Move UserDTO to dto package
- Fix home controller test

minzzzy committed:
chore: Remove cart acceptance test

javajigi commented on src/main/java/codesquad/domain/Cart.java:
```diff
@@ -0,0 +1,127 @@
+    @MapKeyJoinColumn
+    @CollectionTable
+    @Lob
+    private HashMap<Long, Integer> products;
```
HashMap 보다는 map 추천

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -0,0 +1,110 @@
+});
+
+class Cart {
+    constructor() {
```
cart 객체를 만들때 필요한 변수값을 넘겨줘도 되겠죠. 
아래처럼요.

const cart = $('#cart');
new Cart(cart);

javajigi commented on src/main/java/codesquad/domain/Cart.java:
```diff
@@ -0,0 +1,127 @@
+    @DateTimeFormat
+    private Date orderDate;
+
+    private Long productsTotalPrice;
```
이 값은 다른 값들을 이용해 구할 수 있는데 굳이 인스턴스 변수로 구현해야할까?

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    updateProductInCart() {
+        this.cart.addEventListener('click', (e) => {
```
콜백의 몸체가 크면 별도로 분리해서 선언하면 보기 좋아요.

this.cart.addEventListener('click', (e) => { this.minjihandler(e) };

minjihandler() {
  ..//궁시렁궁시렁
}

javajigi commented on src/main/java/codesquad/domain/Cart.java:
```diff
@@ -0,0 +1,127 @@
+
+    private Long productsTotalPrice;
+
+    private Long deliveryCharge;
```
이 값은 다른 값들을 이용해 구할 수 있는데 굳이 인스턴스 변수로 구현해야할까?

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -0,0 +1,110 @@
+            if (target.classList.contains('down')) {
+                const buyCnt = this.buyCntBox(target);
+                if (buyCnt.value <= 1){
+                    alert("수량이 1보다 작을 순 없습니다.");
```
얼럿은 실무에서 안씁니다요.

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -0,0 +1,110 @@
+        });
+    }
+
+    computeTotalPrice() {
```
fetchUpdateProduct와 computeTotalPrice 이것처럼 비슷한 여러개를 만들기보다, 
하나의 메서드로 callback과 url등을  받아서 동작되게 하는 것도 좋을 거 같네요.
중복코드는 늘 안좋아요.

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -0,0 +1,110 @@
+
+    changeProductData(data, target) {
+        const productPrice = target.closest('td').parentNode.querySelector('.productPrice');
+        const sumSalePrice = target.closest('td').parentNode.querySelector('.sumSalePrice');
```
여기서도 중복인데, 컴퓨터에게 똑같은 일을 계속 시키는 셈이에요.

const pNode = target.closest('td').parentNode;
pNode.querySelector(".productPrice');

javajigi commented on src/test/java/codesquad/domain/CartTest.java:
```diff
@@ -0,0 +1,180 @@
+        HashMap<Long, Integer> productIds = new HashMap<>();
+        productIds.put(product.getId(), productNum);
+        return new Cart(productIds);
+    }
```
단위 테스트 👍

javajigi reviewed (changes requested):
객체 설계 및 단위 테스트 구현 잘 했네요. 💯 굳이 불필요한 인스턴스 변수를 추가하지 않으면 더 좋을 것 같아요.

crongro reviewed (approved):
JS 클래스를 좀더 범용성있게 쓸수 있도록 파라미터를 사용해보세요.

지금의 메서드 크기는 적당해요.

어떤 경우이든 코드의 중복을 제거하는 것부터 시도하시면 좋은코드를 만들 수 있어요.

바쁜와중에 수고했네요.

minzzzy committed:
refactor: Rename Dto to DTO

minzzzy committed:
refactor: Remove price fields in Cart

- Rename CartStaticValue to CartValue
- Add discountedPrice method in Cart with test
- Move CartValue to common package
- Change products type to LinkedHashMap for order

minzzzy committed:
refactor: Change fetch manager to promise pattern

- Remove duplicated code

minzzzy committed:
feat: Add error message box

minzzzy committed:
refactor: Change annotation to add date when create and update entity

minzzzy committed:
feat: Add handler method argument resolver of cart session with test

- Create cart if not exist when handling the cart argument
- Rename method of existing cart in session to 'hasCartSession'

minzzzy committed:
refactor: Add cart acceptance test and refactor cart controller back-end

- Change cart page to rest controller
- Add cart acceptance test

minzzzy committed:
refactor: Create method to add product with test

minzzzy committed:
chore: Rename method of getting products list

minzzzy committed:
refactor: Change server side rendering to client side rendering

- Add request method to get cart data

minzzzy committed:
chore: Delete unused variables

minzzzy committed:
refactor: Refactor Baeminchan controller

- Remove CartController and move cart route to Baeminchan controller
- Add Baeminchan controller acceptance test
- Rename cart controller acceptance test class

minzzzy committed:
feat: Add back-end logic for removing a product in cart

- Add deletion logic with acceptance test

minzzzy committed:
feat: Add front-end logic for removing a product in cart

minzzzy committed:
refactor: Refactor javascript files

- Prevent default event when click
- Remove duplicated request code

minzzzy committed:
refactor: Refactor ControllerAdvice for handling error message

- Add error response in ResponseModel
- Move response model domain to common package
- Add Message by using message handling resource
- Simplify UserVerifyException just for using message
- Add ResourceNotFoundException in ControllerAdvice
- Rename ValidationAdvice to ControllerAdvice
- Remove unused validation error and response object
- Add messageSourceAccessor mock in service test
- Change user request format in user acceptance test

minzzzy committed:
refactor: Show error message in front-end

- Create error message function
- Handle and show error messages

minzzzy committed:
refactor: Handle not to exist product by finding id

- Add service and acceptance test

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -45,9 +46,33 @@ class Cart {
             return;
         }
 
+        if (target.classList.contains('delete_cart_item')) {
```
a = document.querySelector("div").classList; 
이렇게 변수로 담아두고 쓰면, 중복도 없어지고 더 빠른코드가 됩니다.
캐시죠

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -63,12 +88,63 @@ class Cart {
             });
     }
 
+    deleteProduct(target) {
```
만약
함수이름은 delete지만 이녀석이  무언가를 반환한다면 반환하는 정보를 메서드명으로 하는게  더 정확해요.

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -63,12 +88,63 @@ class Cart {
+    }
+
+    checkAllBoxSelected() {
+        if ($All('.product_chk').length == this.getCheckBoxArray().filter(e => e.checked).length) {
```
조건문안에 코드가 길면 무엇인지 한참들여다보게 돼요.
작은 함수로 만들고, 조건문안에서 함수를 호출하면 '의도'가 더 쉽게 잘 보여요~

crongro commented on src/main/resources/static/js/cart.js:
```diff
@@ -158,44 +159,16 @@ class Cart {
     }
 
-    requestProductList() {
+    request(path, method) {
```
늘 인자를 많이 쓰는게 함수가 함수답게 돼죠.

ryuneeee replied:
하지만 그래도 인터페이스는 깔끔하고 크게크게! 자주 변하지 않아야 하죠. 🤔

crongro reviewed (approved):
점점 발전하는 모습이 좋네요.

minzzzy commented:
크롱 코드리뷰 감사합니다~~

minzzzy committed:
refactor: Refactor Javascript files

- Allocate classList of target when clicked
- Extract function whether all checkboxes are checked
- Rename function

minzzzy committed:
chore: Change static variable type to primitive type

minzzzy committed:
refactor: Add multiple products deletion mapping controller with test

- Rename delete to deleteProduct
- Add acceptance test for deleting multiple products
- Refactor deletion request in front-end
